compiler/codegen/CCData.hpp

```diff
@@ -0,0 +1,207 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef OMR_CCDATA_INCL
+#define OMR_CCDATA_INCL
+
+#include <map>
+#include <string>
+#include <cstddef>
+
+namespace TR { class Monitor; }
+
+namespace OMR
+{
+
+/**
+ * @class CCData
+ *
+ * @brief This class represents a table that can be used to implement a GOT or TOC or a PLT or similar.
+ *
+ * <TODO: A more detailed discussion about the design here.>
+ */
+class CCData
+   {
+   private:
+      /** \typedef data_t Implementation detail. This type represents the units of the table. Typically bytes, but can be some other data type, as long as it can be default-constructed. */
+      typedef uint8_t data_t;
```

Comment:
It can be changed, but the choice isn't arbitrary, it will have implications. The size of `data_t` represents the addressible units of storage in CCData; anything smaller than `sizeof(data_t)` bytes will be occupy an entire `data_t`, and the minimum alignment for anything stored will be `alignof(data_t)`. For my purposes bytes is sufficient, but I wanted to make sure that a) the code could handle other units and b) if someone wanted to change it in the future they'd only have to make the change in a single place.

It could be a template parameter, but I chose not to because there wasn't a strong motivation to make it easily and conveniently changeable, and so that most of the implementation could live in a single compilation unit. IIRC some of the compiler/linker combos we use will not prune duplicate template instantiations so I don't want to contribute to excessive binary sizes.